Generate Versions JSON for new Version Switcher

cristianjd opened the pull request:
## Why are these changes needed?

In order to move away from the "Read the Docs" version switcher and to the one provided by the PyData theme we must first have a stable JSON of the available versions as described here: https://github.com/google/google-java-format/releases/download/google-java-format-1.7/google-java-format-1.7-all-deps.jar

The adds helpers to custom directives in order to get the list of git tags starting from the first one we display in the switcher then sort in semver order. "Latest" and "master" are added to the top. This is run in the setup fn to generate the JSON in the static folder. After this is deployed we can reference it, add the new picker, and hide the previous one.

## Related issue number

This is part one of a fix for https://github.com/ray-project/ray/issues/46382 and https://github.com/ray-project/ray/issues/46189. The follow up will close these out. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

peytondmurray commented on doc/source/custom_directives.py:
```diff
@@ -1252,6 +1255,45 @@ def pregenerate_example_rsts(
+
+        return []
+    except Exception as e:
+        print(f"Error fetching versions: {e}")
```
I'm wondering if we should just be passing through without issue if we can't generate the JSON needed for the version picker. What happens if there _is_ an error - does the build process complete without a `versions.json`? What happens if we _don't_ wrap in a try/except here - what kind of errors do we get? It might be more informative to catch a specific error rather than any error.

In any case, I think we should use the logging system if we are going to emit a message to the user, rather than a print.

peytondmurray reviewed (changes requested):
Thanks for this - left a few comments.

cristianjd commented on doc/versions.json:
sure I can add that

cristianjd commented on doc/source/custom_directives.py:
```diff
@@ -1252,6 +1255,45 @@ def pregenerate_example_rsts(
+                    versions.append(version)
+        versions.sort(key=Version, reverse=True)
+
+        version_json_data = [
+            {"version": v, "url": generate_version_url(v)} for v in ["latest", "master"]
+        ] + [
+            {"version": f"releases/{v}", "url": generate_version_url(f"releases-{v}")}
+            for v in versions
+        ]
```
yup sure thing. I'm a bit of a python noob. will clean this up

cristianjd commented on doc/source/custom_directives.py:
```diff
@@ -1252,6 +1255,45 @@ def pregenerate_example_rsts(
+def generate_versions_json():
+    try:
+        versions = []
+        tags = subprocess.check_output(["git", "tag"]).decode("utf-8").split()
```
Is there a place to put this such that it only happens when building for real and not locally? I didn't have much luck finding a testing an alternative

peytondmurray replied:
No, this is the place to do it. I'm not sure about the best way to do this, but here are some thoughts:

1. People who fork the ray repo often alias the upstream repo to `upstream`. For me, I can fetch all tags from the upstream repo with `git fetch upstream --tags`
2. If this is being run on a github runner, I believe there's a way to fetch tags in the options for `actions/checkout` since I think that does a shallow clone by default. Probably safest to do something in the github layer
3. I haven't searched for a way to get the upstream repo, but maybe someone else has posted on SO looking for the definitive way to find the root URL for a repository? That way you can find out whether someone has cloned a fork of the repo or whether they've cloned `ray-project/ray`.
4. You might be able to check for the presence of certain branches...but that seems like it might have edge cases that don't work?

can-anyscale commented on doc/source/conf.py:
```diff
@@ -483,6 +484,8 @@ def _autogen_apis(app: sphinx.application.Sphinx):
 
 
 def setup(app):
+    generate_versions_json()
```
i think we can limit this function to run only on the readthedocs environment? Couple of ways to detect the readthedocs environment I can think of
- The $READTHEDOCS_OUTPUT is set and not null
- We can add a `pre_build` step to https://github.com/ray-project/ray/blob/master/.readthedocs.yaml (https://docs.readthedocs.io/en/stable/build-customization.html), and run this function as a `pre_build` step instead of inside `conf.py`?

cristianjd commented:
Updated with the following changes:
- added `versions.json` to `.gitignore`
- updated comment to doc string
- removed unnecessary return and generic error handling
- updated generation with the following changes
    - fetch tags directly from repo via `ls-remote` and sanitize them
    - simplified looping logic to not use list comprehensions
    - ensured that the static path exists before writing to that file
 - condition in the `setup` fn so that this is only run when the read the docs flag is set

can-anyscale reviewed (commented):
Nice, did you test this, the readthedoc build seems red

![](https://media0.giphy.com/media/Od0QRnzwRBYmDU3eEO/giphy.gif?cid=5a38a5a24365m8vgzfklg8k8uw4b1qdrd19xey39m83yi2bv&ep=v1_gifs_search&rid=giphy.gif&ct=g)

aslonnie reviewed (commented):
for @can-anyscale to review

cristianjd commented:
@can-anyscale I confirmed that the JSON is generated correctly with the read the docs build
![Screenshot 2024-08-01 at 3 42 44 PM](https://github.com/user-attachments/assets/a55ee959-c46d-45b1-86a5-61107c8a41d4)

removing the revert now

can-anyscale reviewed (approved):
![](https://media3.giphy.com/media/G5MDBwmdTrVMpuRJix/giphy.gif?cid=5a38a5a2c20xzfcaph6164upo1d8j71udjinfoecnng7opnk&ep=v1_gifs_search&rid=giphy.gif&ct=g)
